app/src/main/java/org/wikipedia/edit/EditSectionActivity.kt

```diff
@@ -106,9 +110,45 @@ class EditSectionActivity : BaseActivity(), ThemeChooserDialog.Callback {
 
     private val requestInsertMedia = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
         if (it.resultCode == InsertMediaActivity.RESULT_INSERT_MEDIA_SUCCESS) {
-            it.data?.let { intent ->
-                binding.editSectionText.inputConnection?.commitText("${intent.getStringExtra(InsertMediaActivity.RESULT_WIKITEXT)}", 1)
+            it.data?.let { data ->
+
+                // pass the resulting data into our own current intent, so that we can pass it back
+                // to the InsertImage workflow if the user navigates back to it.
+                val imageTitle = data.getParcelableExtra<PageTitle>(InsertMediaActivity.EXTRA_IMAGE_TITLE)
+                val imageCaption = data.getStringExtra(InsertMediaActivity.RESULT_IMAGE_CAPTION)
+                val imageAlt = data.getStringExtra(InsertMediaActivity.RESULT_IMAGE_ALT)
+                val imageSize = data.getStringExtra(InsertMediaActivity.RESULT_IMAGE_SIZE)
+                val imageType = data.getStringExtra(InsertMediaActivity.RESULT_IMAGE_TYPE)
+                val imagePos = data.getStringExtra(InsertMediaActivity.RESULT_IMAGE_POS)
+
+                intent.putExtra(InsertMediaActivity.EXTRA_IMAGE_TITLE, imageTitle)
+                intent.putExtra(InsertMediaActivity.RESULT_IMAGE_CAPTION, imageCaption)
+                intent.putExtra(InsertMediaActivity.RESULT_IMAGE_ALT, imageAlt)
+                intent.putExtra(InsertMediaActivity.RESULT_IMAGE_SIZE, imageSize)
+                intent.putExtra(InsertMediaActivity.RESULT_IMAGE_TYPE, imageType)
+                intent.putExtra(InsertMediaActivity.RESULT_IMAGE_POS, imagePos)
+
+                val newWikiText = InsertMediaViewModel.insertImageIntoWikiText(pageTitle.wikiSite.languageCode,
+                    sectionWikitext.orEmpty(), imageTitle?.text.orEmpty(), imageCaption.orEmpty(),
+                    imageAlt.orEmpty(), imageSize.orEmpty(), imageType.orEmpty(), imagePos.orEmpty(),
+                    if (invokeSource == Constants.InvokeSource.EDIT_ADD_IMAGE) 0 else binding.editSectionText.selectionStart,
+                    invokeSource == Constants.InvokeSource.EDIT_ADD_IMAGE,
+                    intent.getBooleanExtra(InsertMediaActivity.EXTRA_ATTEMPT_INSERT_INTO_INFOBOX, false))
+
+                binding.editSectionText.setText(newWikiText.first)
+                intent.putExtra(InsertMediaActivity.EXTRA_INSERTED_INTO_INFOBOX, newWikiText.second)
+
+                // TODO: automatically highlight what was added.
+                // binding.editSectionText.setSelection(cursorPos, cursorPos + newText.length)
```

Comment:
Just curious, is this something that will be implemented in the future?

Reply:
It will be, but the commented-out line doesn't seem to work consistently across devices. We'll need to revisit later.